src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveHandler.cs

```diff
@@ -170,19 +171,30 @@ public async Task<LSP.VSCodeAction> HandleRequestAsync(LSP.VSCodeAction codeActi
             {
                 foreach (var docId in changedDocuments)
                 {
-                    var newDoc = getNewDocumentFunc(docId);
-                    var oldDoc = getOldDocumentFunc(docId);
+                    var newTextDoc = getNewDocumentFunc(docId);
+                    var oldTextDoc = getOldDocumentFunc(docId);
 
-                    Contract.ThrowIfNull(oldDoc);
-                    Contract.ThrowIfNull(newDoc);
+                    Contract.ThrowIfNull(oldTextDoc);
+                    Contract.ThrowIfNull(newTextDoc);
 
-                    var oldText = await oldDoc.GetTextAsync(cancellationToken).ConfigureAwait(false);
-                    var newText = await newDoc.GetTextAsync(cancellationToken).ConfigureAwait(false);
+                    var oldText = await oldTextDoc.GetTextAsync(cancellationToken).ConfigureAwait(false);
 
-                    var textChanges = newText.GetTextChanges(oldText);
+                    IEnumerable<TextChange> textChanges;
+
+                    // Normal documents have a unique method for calculating minimal text edits. If we used the standard 'GetTextChanges'
+                    // method instead, we would get a change that spans the entire document, which we ideally want to avoid.
+                    if (newTextDoc is Document newDoc && oldTextDoc is Document oldDoc)
+                    {
+                        textChanges = await newDoc.GetTextChangesAsync(oldDoc, cancellationToken).ConfigureAwait(false);
+                    }
+                    else
+                    {
+                        var newText = await newTextDoc.GetTextAsync(cancellationToken).ConfigureAwait(false);
+                        textChanges = newText.GetTextChanges(oldText);
+                    }
```

Comment:
we have a IDocumentTextDifferencingService.  can you investigate if we should be using that instead?

Reply:
actually, i'm fairly certain you should use that.  That service will fall bcak to this logic, but can use even better logic in a host that supports it.